From 99d9934ba23e6693e1da5d3c2d32a2b32b455264 Mon Sep 17 00:00:00 2001
From: Neel Patel <nspatel1986@yahoo.com>
Date: Wed, 25 Aug 2021 17:48:55 -0700
Subject: [PATCH] dts/pensando: creating reserved dma memory pool for mnet
 devices

This patch creates a reserved memory region for use by the mnet
instances.  This avoids memory allocation failure due to wider system
memory fragmentation.

Co-authored-by: Neel Patel <neel@pensando.io>
Signed-off-by: David Clear <dac2@pensando.io>
---
 .../boot/dts/pensando/elba-asic-memory.dtsi   | 16 +++++
 arch/arm64/boot/dts/pensando/elba-asic.dts    |  2 +
 arch/arm64/boot/dts/pensando/elba-mdev.dtsi   | 71 +++++++++++++++++++
 arch/arm64/boot/dts/pensando/elba.dtsi        | 41 -----------
 4 files changed, 89 insertions(+), 41 deletions(-)
 create mode 100644 arch/arm64/boot/dts/pensando/elba-asic-memory.dtsi
 create mode 100644 arch/arm64/boot/dts/pensando/elba-mdev.dtsi

diff --git a/arch/arm64/boot/dts/pensando/elba-asic-memory.dtsi b/arch/arm64/boot/dts/pensando/elba-asic-memory.dtsi
new file mode 100644
index 000000000000..6652b42f1a64
--- /dev/null
+++ b/arch/arm64/boot/dts/pensando/elba-asic-memory.dtsi
@@ -0,0 +1,16 @@
+
+/ {
+	reserved-memory {
+		#address-cells = <2>;
+		#size-cells = <2>;
+
+		ranges;
+
+		mnet_reserved: mnet_dma_region {
+			compatible = "shared-dma-pool";
+			no-map;
+			size = <0x0 0x6000000>;
+			alignment = <0x0 0x1000>;
+		};
+	};
+};
diff --git a/arch/arm64/boot/dts/pensando/elba-asic.dts b/arch/arm64/boot/dts/pensando/elba-asic.dts
index d074b1f1574a..06f60284906a 100644
--- a/arch/arm64/boot/dts/pensando/elba-asic.dts
+++ b/arch/arm64/boot/dts/pensando/elba-asic.dts
@@ -5,3 +5,5 @@
 #include "elba-16core.dtsi"
 #include "elba-asic-common.dtsi"
 #include "elba-flash-parts.dtsi"
+#include "elba-asic-memory.dtsi"
+#include "elba-mdev.dtsi"
diff --git a/arch/arm64/boot/dts/pensando/elba-mdev.dtsi b/arch/arm64/boot/dts/pensando/elba-mdev.dtsi
new file mode 100644
index 000000000000..a728a7625f41
--- /dev/null
+++ b/arch/arm64/boot/dts/pensando/elba-mdev.dtsi
@@ -0,0 +1,71 @@
+/ {
+	dma-coherent;
+
+	mnet0: mnet0 {
+		compatible = "pensando,mnet";
+		msi-parent = <&its 0x0>;
+		memory-region = <&mnet_reserved>;
+	};
+	mnet1: mnet1 {
+		compatible = "pensando,mnet";
+		msi-parent = <&its 0x1>;
+		memory-region = <&mnet_reserved>;
+	};
+	mnet2: mnet2 {
+		compatible = "pensando,mnet";
+		msi-parent = <&its 0x2>;
+		memory-region = <&mnet_reserved>;
+	};
+	mnet3: mnet3 {
+		compatible = "pensando,mnet";
+		msi-parent = <&its 0x3>;
+		memory-region = <&mnet_reserved>;
+	};
+	mnet4: mnet4 {
+		compatible = "pensando,mnet";
+		msi-parent = <&its 0x4>;
+		memory-region = <&mnet_reserved>;
+	};
+	mnet5: mnet5 {
+		compatible = "pensando,mnet";
+		msi-parent = <&its 0x5>;
+		memory-region = <&mnet_reserved>;
+	};
+	mnet6: mnet6 {
+		compatible = "pensando,mnet";
+		msi-parent = <&its 0x6>;
+		memory-region = <&mnet_reserved>;
+	};
+	mnet7: mnet7 {
+		compatible = "pensando,mnet";
+		msi-parent = <&its 0x7>;
+		memory-region = <&mnet_reserved>;
+	};
+	mnet8: mnet8 {
+		compatible = "pensando,mnet";
+		msi-parent = <&its 0x8>;
+		memory-region = <&mnet_reserved>;
+	};
+	mnet9: mnet9 {
+		compatible = "pensando,mnet";
+		msi-parent = <&its 0x9>;
+		memory-region = <&mnet_reserved>;
+	};
+
+	mcrypt0: mcrypt0 {
+		compatible = "pensando,mcrypt";
+		msi-parent = <&its 0xb>;
+	};
+	mcrypt1: mcrypt1 {
+		compatible = "pensando,mcrypt";
+		msi-parent = <&its 0xc>;
+	};
+	mcrypt2: mcrypt2 {
+		compatible = "pensando,mcrypt";
+		msi-parent = <&its 0xd>;
+	};
+	mcrypt3: mcrypt3 {
+		compatible = "pensando,mcrypt";
+		msi-parent = <&its 0xe>;
+	};
+};
diff --git a/arch/arm64/boot/dts/pensando/elba.dtsi b/arch/arm64/boot/dts/pensando/elba.dtsi
index 45b573e44c86..96af62b7e5e5 100644
--- a/arch/arm64/boot/dts/pensando/elba.dtsi
+++ b/arch/arm64/boot/dts/pensando/elba.dtsi
@@ -163,47 +163,6 @@
 			status = "disabled";
 		};
 
-		mnet0: mnet0 {
-			compatible = "pensando,mnet";
-			msi-parent = <&its 0x0>;
-		};
-		mnet1: mnet1 {
-			compatible = "pensando,mnet";
-			msi-parent = <&its 0x1>;
-		};
-		mnet2: mnet2 {
-			compatible = "pensando,mnet";
-			msi-parent = <&its 0x2>;
-		};
-		mnet3: mnet3 {
-			compatible = "pensando,mnet";
-			msi-parent = <&its 0x3>;
-		};
-		mnet4: mnet4 {
-			compatible = "pensando,mnet";
-			msi-parent = <&its 0x4>;
-		};
-		mnet5: mnet5 {
-			compatible = "pensando,mnet";
-			msi-parent = <&its 0x5>;
-		};
-		mnet6: mnet6 {
-			compatible = "pensando,mnet";
-			msi-parent = <&its 0x6>;
-		};
-		mnet7: mnet7 {
-			compatible = "pensando,mnet";
-			msi-parent = <&its 0x7>;
-		};
-		mnet8: mnet8 {
-			compatible = "pensando,mnet";
-			msi-parent = <&its 0x8>;
-		};
-		mnet9: mnet9 {
-			compatible = "pensando,mnet";
-			msi-parent = <&its 0x9>;
-		};
-
 		/* UIO device using interrupt line PCIEMAC */
 		pciemac {
 			#address-cells = <2>;
-- 
2.17.1

